pageserver/src/tenant/timeline.rs

```diff
@@ -4759,31 +4705,176 @@ impl Deref for TimelineWriter<'_> {
     }
 }
 
+impl Drop for TimelineWriter<'_> {
+    fn drop(&mut self) {
+        let state = &mut *self.write_guard;
+        *state = None;
```

Comment:
```suggestion
        self.write_guard.take();
```

This would probably work just as well.

Reply:
Done in https://github.com/neondatabase/neon/pull/6661/commits/0b31fec1f6a5a8600ab3b0225466236e6f459f79. I also replaced some derefs with `replace`.

Reply:
> So the mutex guard is like a scratch space?

Yes. The guard holds the state for the current writer and `None` if there's no writer. It's sort of arbitrary. Holding that state in the writer would have worked just as well.

Reply:
I like holding the state inside the guard because it makes it clear the guard needs to be held.
A brief comment explaining this, i.e., that the state doesn't outlive the TimelineWriter instance that created it, might be helpful for future readers.